xtask/src/themekeyscheck.rs

```diff
@@ -0,0 +1,430 @@
+            match Query::new(language, &query_text) {
+                Err(reason) => {
+                    return Err(format!(
+                        "Failed to parse {} queries for {}: {}",
+                        highlight_queries_file, language_name, reason
+                    ))
+                }
+                Ok(query) => {
+                    let filtered_names = query.capture_names().iter().filter(|cn| {
+                        !(cn.contains("._")
+                            || cn.contains("__")
+                            || cn.contains("_.")
+                            || cn.starts_with('_'))
+                    });
+                    for capture_name in filtered_names {
+                        highlights_queries
+                            .entry(capture_name.clone())
+                            .or_default()
+                            .push(language_name.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    let mut has_error = false;
+    for (hq, langs) in highlights_queries {
+        if !is_known_key(&hq) {
+            eprintln!(
+                "'{}' is not a static theme key (seen in languages: {:?}) ",
+                hq, langs
+            );
+            has_error = true;
+        }
+    }
+
+    if has_error {
+        Err("Theme key are incomplete, add missing keys to the list in `themekeyscheck.rs`".into())
+    } else {
+        println!("All theme keys found in highlight queries are known !");
+        Ok(())
+    }
+}
+
+/// Generate the doc table for `book/src/theme.md`
+pub fn theme_keys_doc() -> String {
+    fn max_size(lengths: &[(&'static str, &'static str)]) -> (usize, usize) {
+        lengths.iter().fold((0, 0), |(km, nm), (kl, nl)| {
+            (km.max(kl.len()), nm.max(nl.len()))
+        })
+    }
+
+    let (keys_w, notes_w) = max_size(KEY_DOCS);
+
+    let mut table = String::from(
+        "<!-- DO NOT MODIFY MANUALLY: Document generated by xtask::themekeyscheck::theme_keys_doc() using `cargo xtask docgen` -->",
+    );
+
+    let write_table_header = |table: &mut String, keys_type: &'static str| {
+        table.push('\n');
+        ::std::writeln!(
+            table,
+            "| {k:kw$} | {n:nw$} |",
+            k = keys_type,
+            kw = keys_w,
+            n = "Notes",
+            nw = notes_w
+        )
+        .unwrap();
+        ::std::writeln!(
+            table,
+            "| {k:kw$} | {n:nw$} |",
+            k = "---",
+            kw = keys_w,
+            n = "---",
+            nw = notes_w
+        )
+        .unwrap();
+    };
+
+    write_table_header(&mut table, "UI-only Keys");
+
+    for &(key, note) in KEY_DOCS {
+        if key == FIRST_MIXED_KEY {
+            write_table_header(&mut table, "Mixed Keys (UI & languages)");
+        } else if key == FIRST_LANGUAGE_KEY {
+            write_table_header(&mut table, "Language-only Keys");
+        }
+
+        ::std::writeln!(
```

Comment:
minor style nit but it's more common in the helix codebase not to use the global path, just using `writeln!`

Reply:
Done